src/hardware/serialport/serialmouse.cpp

```diff
@@ -28,452 +28,551 @@
-            name           = "Microsoft, 2 buttons";
-            byte_len       = 7;
-            has_3rd_button = false;
-            has_wheel      = false;
-            break;
-        case MouseType::Logitech:
-            name           = "Logitech, 3 buttons";
-            byte_len       = 7;
-            has_3rd_button = true;
-            has_wheel      = false;
-            break;
-        case MouseType::Wheel:
-            name           = "wheel, 3 buttons";
-            byte_len       = 7;
-            has_3rd_button = true;
-            has_wheel      = true;
-            break;
-        case MouseType::MouseSystems:
-            name           = "Mouse Systems, 3 buttons";
-            byte_len       = 8;
-            has_3rd_button = true;
-            has_wheel      = false;
-            break;
-        default: LogUnimplemented(); break;
-        }
-
-        if (name)
-            LOG_MSG("MOUSE (COM%d): %s", port_num, name);
-    }
+	std::string option;
+	if (cmd->FindStringBegin("rate:", option, false))
+		LOG_WARNING("MOUSE (COM%d): Deprecated option 'rate:' - ignored",
+		            port_num);
+
+	const bool found_deprecated = cmd->FindStringBegin("type:", option, false);
+
+	if (found_deprecated) {
+		LOG_WARNING("MOUSE (COM%d): Deprecated option 'type:'", port_num);
+
+		if (option == "2btn") {
+			param_model    = MouseModelCOM::Microsoft;
+			param_auto_msm = false;
+		} else if (option == "2btn+msm") {
+			param_model    = MouseModelCOM::Microsoft;
+			param_auto_msm = true;
+		} else if (option == "3btn") {
+			param_model    = MouseModelCOM::Logitech;
+			param_auto_msm = false;
+		} else if (option == "3btn+msm") {
+			param_model    = MouseModelCOM::Logitech;
+			param_auto_msm = true;
+		} else if (option == "wheel") {
+			param_model    = MouseModelCOM::Wheel;
+			param_auto_msm = false;
+		} else if (option == "wheel+msm") {
+			param_model    = MouseModelCOM::Wheel;
+			param_auto_msm = true;
+		} else if (option == "msm") {
+			param_model    = MouseModelCOM::MouseSystems;
+			param_auto_msm = false;
+		} else {
+			LOG_ERR("MOUSE (COM%d): Invalid type '%s'",
+			        port_num,
+			        option.c_str());
+			return;
+		}
+	}
+}
+
+void CSerialMouse::BoostRate(const uint16_t rate_hz)
+{
+	if (!rate_hz || model == MouseModelCOM::NoMouse) {
+		rate_coeff = 1.0f;
+		return;
+	}
+
+	// Estimate current sampling rate, as precisely as possible
+	auto estimate = [](const uint16_t bauds,
+	                   const uint8_t byte_len,
+	                   const MouseModelCOM model) {
+		// In addition to byte_len, the mouse has to send
+		// 3 more bits per each byte: start, parity, stop
+
+		if (model == MouseModelCOM::Microsoft ||
+		    model == MouseModelCOM::Logitech || model == MouseModelCOM::Wheel)
+			// Microsoft-style protocol
+			// single movement needs exactly 3 bytes to be reported
+			return bauds / (static_cast<float>(byte_len + 3) * 3.0f);
+		else if (model == MouseModelCOM::MouseSystems)
```

Comment:
But `model` here is an argument passed into the `estimate(...)` lamba.

Reply:
Changed, captured `this`.